L64-python-runtime-proto-parsing.md

```diff
@@ -0,0 +1,99 @@
+
+Open source users are aware of this gap in the ecosystem and are actively looking for ways to fill it. Many have settled on [protocol buffer monorepos](https://medium.com/namely-labs/how-we-build-grpc-services-at-namely-52a3ae9e7c35) as a solution to the problem, wherein all `.proto` files for an organization are placed in a single source code repository and included by all other repositories as a submodule. But even this is not a complete solution. In addition, some mechanism must be put in place for the repositories housing the client and server code to retrieve the desired protocol buffer and generate code for the target language.
+
+The protocol compiler paired with Google's build system also means that the average engineer never has to manually invoke `protoc`. Instead, when an update is made to a `.proto` file and some code file references it, the code for the protocol buffer is regenerated without any manual intervention on the part of the engineer. Compare that to today's workflow for gRPC users:
+
+ 1. Update .proto file.
+ 2. Manually regenerate code. (remembering how to use all of the CLI flags)
+ 3. Make necessary corresponding updates to code using the protocol buffer.
+ 4. Rerun.
+
+It's easy for several of those steps to slip one's mind while developing. Moreover, figuring out how to invoke `protoc` in a way that meshes with your imports can be quite difficult. Python developers in particular are unused to build-time steps such as these; it is much more common to perform them at runtime.
+
+### Related Proposals
+
+* [Node.js gRPC+Protobuf.js Library](https://github.com/grpc/proposal/blob/master/L23-node-protobufjs-library.md)
+
+
+## Proposal
+
+Whereas today, users import protobuf code by invoking `protoc` to generate a file called `foo_pb2.py` and including the following line of code:
+
+```python
+import foo_pb2
+import foo_pb2_grpc
+```
+
+They will now also have the option of completely skipping the manual protoc invocation and instead writing
+
+```python
+protos = grpc.protos('foo.proto')
+services = grpc.services('foo.proto')
+```
+
+These two new functions return the same module objects as the `import foo_pb2` and `import foo_pb2_grpc` statements. In order to maintain interoperability with any proto-backed modules loaded in the same process, after these functions are invoked, `import foo_pb2` and `import foo_pb2_grpc` will be no-ops. That is, a side-effect of calling `grpc.protos` and `grpc.services` is insertion of the returned modules into the per-process cache. This ensures that regardless of whether the application calls `grpc.protos('foo.proto')` or `import foo_pb2` or in which order, only a single version of the module will ever be loaded into the process. This avoids situations in which interoperability breaks due to two modules expecting the same protobuf-level message type, but expecting two *different* Python-level `Message` classes, one backed by a `_pb2.py` file and one backed by a `.proto` file.
```

Comment:
Want to check if I read it correct. In order to avoid duplication, the `grpc.protos('foo.proto')` registers a global module with name "foo_pb2"? So, the following `import foo_pb2` will import the generated module instead of trying to load from disk?

Reply:
Almost. The canonical name of the module imported by `grpc.protos('foo.proto')` will be `foo_pb2`, so the subsequent `import foo_pb2` operation won't result in an "import" at all. It will just make `foo_pb2` point to the module loaded from the file "foo.proto". That's a very nitpicky difference from how you stated it though.